uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-infinispan/src/main/java/org/uberfire/ext/metadata/backend/infinispan/proto/KObjectMarshaller.java

```diff
@@ -0,0 +1,284 @@
+                                                              CLUSTER_ID,
+                                                              SEGMENT_ID,
+                                                              MetaObject.META_OBJECT_KEY,
+                                                              MetaObject.META_OBJECT_FULL_TEXT);
+
+    public KObjectMarshaller(String typeName) {
+
+        this.typeName = typeName;
+    }
+
+    @Override
+    public KObjectImpl readFrom(ProtoStreamReader protoStreamReader) throws IOException {
+
+        Descriptor descriptor = protoStreamReader.getSerializationContext().getMessageDescriptor(this.getTypeName());
+
+        List<KProperty<?>> properties = descriptor.getFields()
+                .stream()
+                .filter(fieldDescriptor -> !isMainAttribute(fieldDescriptor))
+                .filter(fieldDescriptor ->
+                                isExtension(descriptor.getName())
+                )
+                .map(field -> (KProperty<?>) new KPropertyImpl(toKPropertyFormat(field.getName()),
+                                                               this.read(field,
+                                                                         protoStreamReader),
+                                                               false)
+                )
+                .collect(toList());
+
+        String id = protoStreamReader.readString(MetaObject.META_OBJECT_ID);
+        String type = protoStreamReader.readString(MetaObject.META_OBJECT_TYPE);
+        String clusterId = protoStreamReader.readString(CLUSTER_ID);
+        String segmentId = protoStreamReader.readString(SEGMENT_ID);
+        String key = protoStreamReader.readString(MetaObject.META_OBJECT_KEY);
+        String fullText = protoStreamReader.readString(MetaObject.META_OBJECT_FULL_TEXT);
+
+        return new KObjectImpl(id,
+                               type,
+                               clusterId,
+                               segmentId,
+                               key,
+                               properties,
+                               !StringUtils.isEmpty(fullText));
+    }
+
+    @Override
+    public void writeTo(ProtoStreamWriter protoStreamWriter,
+                        KObject kObject) throws IOException {
+
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_ID,
+                                      kObject.getId());
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_TYPE,
+                                      kObject.getType().getName());
+        protoStreamWriter.writeString(CLUSTER_ID,
+                                      kObject.getClusterId());
+        protoStreamWriter.writeString(SEGMENT_ID,
+                                      kObject.getSegmentId());
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_KEY,
+                                      kObject.getKey());
+
+        Map<String, KProperty<?>> props = new HashMap<>();
```

Comment:
This is potentially where the fields are written out of order, as the ```HashMap``` doesn't have predictable iteration order.